File: ocaml/tea/lib-tea/tea.ml
(*
   Cmd represents impure operations that will be executed by a runtime.

   You "command" the runtime to execute those impure actions.
*)
type 'action cmd = Cmd of (unit -> 'action) list

let cmd_none : 'action cmd = Cmd []

let cmd_batch (cmds : 'action cmd list) : 'action cmd =
  Cmd (List.concat_map (fun (Cmd actions) -> actions) cmds)
;;

let run_cmd update_f model (Cmd actions) =
  List.fold_left
    (fun m a -> fst (update_f m a))
    model
    (List.map (fun action -> action ()) actions)
;;

(* USER AVAILABLE COMMANDS *)

type tea_time = Tea_time of float

let to_tea_time time = Tea_time time

let show_tea_time (Tea_time time) =
  let tm = Unix.localtime time in
  let msec = int_of_float (1000.0 *. (time -. floor time)) in
  Printf.sprintf
    "%04d-%02d-%02d %02d:%02d:%02d.%03d"
    (tm.tm_year + 1900)
    (tm.tm_mon + 1)
    tm.tm_mday
    tm.tm_hour
    tm.tm_min
    tm.tm_sec
    msec
;;

let request_time to_action =
  Cmd [ (fun () -> to_action (to_tea_time (Unix.gettimeofday ()))) ]
;;

let request_random_number to_action = Cmd [ (fun () -> to_action (Random.int 100 + 1)) ]

(* RENDERING *)

type console_ui =
  | Line of string
  | BlankLine
  | Section of console_ui list
  | Divider of char * int

let clear_screen () = print_string "\027[2J\027[H"

let rec render_console_ui ui =
  ()
  ; clear_screen ()
  ; _render ui

and _render = function
  | Line str -> print_endline str
  | BlankLine -> print_endline ""
  | Section uis -> List.iter _render uis
  | Divider (char, n) -> print_endline (String.make n char)
;;

(* RUNTIME *)

type ('model, 'action) program =
  { init : 'model * 'action cmd
  ; update : 'model -> 'action -> 'model * 'action cmd
  ; view : 'model -> console_ui
  ; input_to_action : char -> 'action option
  }

let setup_console () = ignore @@ Sys.command "stty cbreak -echo"
let restore_console () = ignore @@ Sys.command "stty -cbreak echo"

let quit () =
  ()
  ; restore_console ()
  ; print_endline "\nGoodbye!"
;;

let rec run_loop program model =
  let view_f = program.view in
  let input_to_action_f = program.input_to_action in
  let console_ui = view_f model in
  render_console_ui console_ui
  ; ()
  ; let input = input_char stdin in
    if input = 'q' then
      quit ()
    else (
      match input_to_action_f input with
      | None -> run_loop program model
      | Some action ->
        let (new_model, new_cmd) = program.update model action in
        let final_model = run_cmd program.update new_model new_cmd in
        run_loop program final_model
    )
;;

let start program =
  let (init_model, init_cmd) = program.init in
  let new_model = run_cmd program.update init_model init_cmd in
  ()
  ; setup_console ()
  ; try run_loop program new_model with
    | e ->
      ()
      ; restore_console ()
      ; raise e
;;

File: haskell/src/TEA.hs
module TEA
    ( start
    , Program
        ( MkProgram
        , cInit
        , cUpdate
        , cView
        , cInputToAction
        )
    , Cmd
    , cmdNone
    , cmdBatch
    , TeaTime
    , ConsoleUI
        ( Line
        , BlankLine
        , Section
        , Divider
        )
    , requestTime
    , requestRandomNumber
    ) where

import Data.Time.Clock (UTCTime)
import qualified Data.Time.Clock as DTC
import System.IO
    ( BufferMode (NoBuffering)
    , hSetBuffering
    , hSetEcho
    , stdin
    )
import System.Random (Random, randomRIO)

import Prelude hiding (init)

-------------------------------------------------------------------------------
--
-- RENDERING
--
-------------------------------------------------------------------------------

data ConsoleUI
    = Line String
    | BlankLine
    | Section [ConsoleUI]
    | Divider Char Int

clearScreen :: IO ()
clearScreen = putStr "\ESC[2J\ESC[H"

-- renderConsoleUI :: ConsoleUI -> IO ()
renderConsoleUI :: ConsoleUI -> IO ()
renderConsoleUI ui =
    clearScreen >>= \() ->
        _render ui

-- _render :: ConsoleUI -> IO ()
_render :: ConsoleUI -> IO ()
_render ui = case ui of
    Line str -> putStrLn str
    BlankLine -> putStrLn ""
    Section uis -> mapM_ _render uis
    Divider char n -> putStrLn (replicate n char)

-------------------------------------------------------------------------------
--
-- RUNTIME
--
-------------------------------------------------------------------------------

data Program model action
    = MkProgram
    { cInit :: (model, Cmd action)
    , cUpdate :: model -> action -> (model, Cmd action)
    , cView :: model -> ConsoleUI
    , cInputToAction :: Char -> Maybe action
    }

setupConsole :: IO ()
setupConsole = do
    hSetBuffering stdin NoBuffering -- Don't buffer key inputs
    hSetEcho stdin False -- Don't echo characters

-- start :: IO ()
start :: Program model action -> IO ()
start cfg = do
    setupConsole
    let (initModel, initCmd) = cInit cfg
    newModel <- runCmd (cUpdate cfg) initModel initCmd
    runLoop cfg newModel

runLoop :: Program model action -> model -> IO ()
runLoop cfg model = do
    let viewF = cView cfg
    let inputToActionF = cInputToAction cfg
    let consoleUI = viewF model :: ConsoleUI
    renderConsoleUI consoleUI

    input <- getChar
    if input == 'q'
        then
            putStrLn "\nGoodbye!"
        else do
            case inputToActionF input of
                Nothing ->
                    runLoop cfg model
                Just action -> do
                    let (newModel, newCmd) = cUpdate cfg model action
                    runLoop cfg =<< runCmd (cUpdate cfg) newModel newCmd

-------------------------------------------------------------------------------
--
-- COMMAND INTERNALS
--
-------------------------------------------------------------------------------

-- Cmd represents side effects that will be executed by a runtime
-- You "command" the runtime to perform these actions
newtype Cmd action = Cmd [IO action]

-- Helper functions to create commands
cmdNone :: Cmd action
cmdNone = Cmd []

cmdBatch :: [Cmd action] -> Cmd action
cmdBatch cmds = Cmd (concatMap (\(Cmd ios) -> ios) cmds)

runCmd :: (model -> action -> (model, Cmd action)) -> model -> Cmd action -> IO model
runCmd updateF model (Cmd ios) =
    foldl (applyUpdate updateF) model <$> sequence ios

applyUpdate :: (model -> action -> (model, Cmd action)) -> model -> action -> model
applyUpdate updateF model action = fst $ updateF model action

-------------------------------------------------------------------------------
--
-- USER AVAILABLE COMMANDS
--
-------------------------------------------------------------------------------

requestRandomNumber :: (Random a, Num a) => (a -> action) -> Cmd action
requestRandomNumber toAction = Cmd [toAction <$> randomRIO (1, 100)]

newtype TeaTime = TeaTime UTCTime

toTeaTime :: UTCTime -> TeaTime
toTeaTime = TeaTime

instance Show TeaTime where
    show (TeaTime time) = show time

requestTime :: (TeaTime -> action) -> Cmd action
requestTime toAction = Cmd [toAction . toTeaTime <$> DTC.getCurrentTime]